pkg/apis/apps/v1alpha1/workloadrebalancer_types.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2024 The Karmada Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+const (
+	// ResourceKindWorkloadRebalancer is kind name of WorkloadRebalancer.
+	ResourceKindWorkloadRebalancer = "WorkloadRebalancer"
+	// ResourceSingularWorkloadRebalancer is singular name of WorkloadRebalancer.
+	ResourceSingularWorkloadRebalancer = "workloadrebalancer"
+	// ResourcePluralWorkloadRebalancer is kind plural name of WorkloadRebalancer.
+	ResourcePluralWorkloadRebalancer = "workloadrebalancers"
+	// ResourceNamespaceScopedWorkloadRebalancer indicates if WorkloadRebalancer is NamespaceScoped.
+	ResourceNamespaceScopedWorkloadRebalancer = false
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:resource:path=workloadrebalancers,scope="Cluster",shortName=wr,categories={karmada-io}
+// +kubebuilder:subresource:status
+// +kubebuilder:storageversion
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// WorkloadRebalancer represents the desired behavior and status of a job which can enforces a rescheduling.
+//
+// Notes: make sure the clocks of controller-manager and scheduler are synchronized when using this API.
+type WorkloadRebalancer struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec represents the specification of the desired behavior of WorkloadRebalancer.
+	// +required
+	Spec WorkloadRebalancerSpec `json:"spec"`
+
+	// Status represents the status of WorkloadRebalancer.
+	// +optional
+	Status WorkloadRebalancerStatus `json:"status,omitempty"`
+}
+
+// WorkloadRebalancerSpec represents the specification of the desired behavior of Reschedule.
+//
+// Notes: this API represents a one-time execution process, once the object is created, the execution process begins,
+// and it will not respond to any modification of the spec field.
+type WorkloadRebalancerSpec struct {
+	// Workloads used to specify the list of expected resource.
+	// Nil or empty list is not allowed.
+	// +kubebuilder:validation:MinItems=1
+	// +required
+	Workloads []Workload `json:"workloads"`
+}
+
+// Workload the expected resource.
+type Workload struct {
```

Comment:
```suggestion
type ObjectReference struct {
```
As
1. the `workload` will probably be reserved for Karmada workloads type.
2. the `workload` can not used to represent PP/CPP in the future.

Reply:
This struct has `apiVersion` and `kind` field, which is not required for future PP/CPP.
 
May be the struct for representing PP/CPP be other field?